File: packages/coding-agent/src/modes/print-mode.ts
/**
 * Print mode (single-shot): Send prompts, output result, exit.
 *
 * Used for:
 * - `pi -p "prompt"` - text output
 * - `pi --mode json "prompt"` - JSON event stream
 */

import type { AssistantMessage, ImageContent } from "@earendil-works/pi-ai";
import type { AgentSessionRuntime } from "../core/agent-session-runtime.ts";
import { flushRawStdout, waitForRawStdoutBackpressure, writeRawStdout } from "../core/output-guard.ts";
import { killTrackedDetachedChildren } from "../utils/shell.ts";
import { toJsonEvent } from "./json-event.ts";

/**
 * Options for print mode.
 */
export interface PrintModeOptions {
	/** Output mode: "text" for final response only, "json" for all events */
	mode: "text" | "json";
	/** Array of additional prompts to send after initialMessage */
	messages?: string[];
	/** First message to send (may contain @file content) */
	initialMessage?: string;
	/** Images to attach to the initial message */
	initialImages?: ImageContent[];
}

/**
 * Run in print (single-shot) mode.
 * Sends prompts to the agent and outputs the result.
 */
export async function runPrintMode(runtimeHost: AgentSessionRuntime, options: PrintModeOptions): Promise<number> {
	const { mode, messages = [], initialMessage, initialImages } = options;
	let exitCode = 0;
	let session = runtimeHost.session;
	let unsubscribe: (() => void) | undefined;
	let unsubscribeBackpressure: (() => void) | undefined;
	let disposed = false;
	const signalCleanupHandlers: Array<() => void> = [];

	const disposeRuntime = async (): Promise<void> => {
		if (disposed) return;
		disposed = true;
		unsubscribe?.();
		unsubscribeBackpressure?.();
		await runtimeHost.dispose();
	};

	const registerSignalHandlers = (): void => {
		const signals: NodeJS.Signals[] = ["SIGTERM"];
		if (process.platform !== "win32") {
			signals.push("SIGHUP");
		}

		for (const signal of signals) {
			const handler = () => {
				killTrackedDetachedChildren();
				void disposeRuntime().finally(() => {
					process.exit(signal === "SIGHUP" ? 129 : 143);
				});
			};
			process.on(signal, handler);
			signalCleanupHandlers.push(() => process.off(signal, handler));
		}
	};

	registerSignalHandlers();

	runtimeHost.setRebindSession(async () => {
		await rebindSession();
	});

	const rebindSession = async (): Promise<void> => {
		session = runtimeHost.session;
		await session.bindExtensions({
			mode: mode === "json" ? "json" : "print",
			commandContextActions: {
				waitForIdle: () => session.waitForIdle(),
				newSession: async (newSessionOptions) => runtimeHost.newSession(newSessionOptions),
				fork: async (entryId, forkOptions) => {
					const result = await runtimeHost.fork(entryId, forkOptions);
					return { cancelled: result.cancelled };
				},
				navigateTree: async (targetId, navigateOptions) => {
					const result = await session.navigateTree(targetId, {
						summarize: navigateOptions?.summarize,
						customInstructions: navigateOptions?.customInstructions,
						replaceInstructions: navigateOptions?.replaceInstructions,
						label: navigateOptions?.label,
					});
					return { cancelled: result.cancelled };
				},
				switchSession: async (sessionPath, switchOptions) => {
					return runtimeHost.switchSession(sessionPath, switchOptions);
				},
				reload: async () => {
					await session.reload();
				},
			},
			onError: (err) => {
				console.error(`Extension error (${err.extensionPath}): ${err.error}`);
			},
		});

		unsubscribe?.();
		unsubscribeBackpressure?.();
		unsubscribe = session.subscribe((event) => {
			if (mode === "json") {
				writeRawStdout(`${JSON.stringify(toJsonEvent(event))}\n`);
			}
		});
		unsubscribeBackpressure =
			mode === "json"
				? session.agent.subscribe(async () => {
						await waitForRawStdoutBackpressure();
					})
				: undefined;
	};

	try {
		if (mode === "json") {
			const header = session.sessionManager.getHeader();
			if (header) {
				writeRawStdout(`${JSON.stringify(header)}\n`);
			}
		}

		await rebindSession();

		if (initialMessage) {
			await session.prompt(initialMessage, { images: initialImages });
		}

		for (const message of messages) {
			await session.prompt(message);
		}

		if (mode === "text") {
			const state = session.state;
			// Walk BACKWARDS to the last assistant message that carries text.
			// A long agentic run can legitimately end on a tool-call-only
			// assistant message, or on a non-assistant entry (compaction,
			// extension injection) — the previous behaviour printed "the
			// literal last entry or nothing", which emitted ZERO bytes with
			// exit 0 on exactly those tails, silently losing the run's report.
			// An error/abort on the terminal assistant message still reports
			// as an error; a session with no printable text anywhere is a
			// FAILED print, not a silent success.
			let printedText = false;
			let reportedError = false;
			for (let i = state.messages.length - 1; i >= 0; i--) {
				const candidate = state.messages[i];
				if (candidate?.role !== "assistant") continue;
				const assistantMsg = candidate as AssistantMessage;
				if (assistantMsg.stopReason === "error" || assistantMsg.stopReason === "aborted") {
					console.error(assistantMsg.errorMessage || `Request ${assistantMsg.stopReason}`);
					exitCode = 1;
					reportedError = true;
					break;
				}
				const textParts = assistantMsg.content.filter((content) => content.type === "text");
				if (textParts.length === 0) continue;
				for (const content of textParts) {
					writeRawStdout(`${content.text}\n`);
				}
				printedText = true;
				break;
			}
			if (!printedText && !reportedError) {
				console.error(
					"pi: print mode produced no text output (no assistant message with text content in the session)",
				);
				exitCode = 1;
			}
		}

		return exitCode;
	} catch (error: unknown) {
		console.error(error instanceof Error ? error.message : String(error));
		return 1;
	} finally {
		for (const cleanup of signalCleanupHandlers) {
			cleanup();
		}
		await disposeRuntime();
		await flushRawStdout();
	}
}

File: python/src/pi_mono/coding_agent/modes/print_mode.py
"""Print mode (single-shot): send prompts, output result, exit."""

from __future__ import annotations

import json
import signal
import sys
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any, Literal

from pi_mono.ai.types import ImageContent
from pi_mono.coding_agent.core.agent_session import AgentSessionRuntime, PromptOptions
from pi_mono.coding_agent.core.auth_guidance import (
    format_api_error_message,
    format_no_api_key_found_message,
    format_no_model_selected_message,
)
from pi_mono.core.output_guard import flush_raw_stdout, write_raw_stdout

PrintOutputMode = Literal["text", "json"]


@dataclass
class PrintModeOptions:
    mode: PrintOutputMode
    messages: list[str] | None = None
    initial_message: str | None = None
    initial_images: list[ImageContent] | None = None


async def run_print_mode(runtime_host: AgentSessionRuntime, options: PrintModeOptions) -> int:
    mode = options.mode
    messages = options.messages or []
    initial_message = options.initial_message
    initial_images = options.initial_images
    exit_code = 0
    session = runtime_host.session
    unsubscribe: Callable[[], None] | None = None
    disposed = False
    signal_handlers: list[tuple[int, Any]] = []

    async def dispose_runtime() -> None:
        nonlocal disposed
        if disposed:
            return
        disposed = True
        if unsubscribe is not None:
            unsubscribe()
        await runtime_host.dispose()

    def register_signal_handlers() -> None:
        for signum in (signal.SIGTERM, signal.SIGHUP):
            try:
                previous = signal.getsignal(signum)

                def handler(_signum: int, _frame: object | None, _previous: Any = previous) -> None:
                    import asyncio

                    async def _exit() -> None:
                        await dispose_runtime()
                        raise SystemExit(129 if _signum == signal.SIGHUP else 143)

                    asyncio.create_task(_exit())

                signal.signal(signum, handler)
                signal_handlers.append((signum, previous))
            except (AttributeError, ValueError, OSError):
                pass

    def unregister_signal_handlers() -> None:
        for signum, previous in signal_handlers:
            try:
                signal.signal(signum, previous)
            except (AttributeError, ValueError, OSError):
                pass
        signal_handlers.clear()

    async def rebind_session() -> None:
        nonlocal session, unsubscribe
        session = runtime_host.session
        await session.bind_extensions()

        if unsubscribe is not None:
            unsubscribe()

        def on_event(event: dict[str, object]) -> None:
            if mode == "json":
                write_raw_stdout(f"{json.dumps(event)}\n")

        unsubscribe = session.subscribe(on_event)

    register_signal_handlers()
    runtime_host.set_rebind_session(rebind_session)

    try:
        if mode == "json":
            header = session.session_manager.get_header()
            if header:
                write_raw_stdout(f"{json.dumps(header)}\n")

        await rebind_session()

        if not session.model or session.model.get("id") in (None, "unknown"):
            print(format_no_model_selected_message(), file=sys.stderr)
            return 1
        auth = await session.model_registry.get_api_key_and_headers(session.model)
        if not auth.get("ok") or not auth.get("apiKey"):
            print(
                auth.get("error")
                or format_no_api_key_found_message(session.model.get("provider", "unknown")),
                file=sys.stderr,
            )
            return 1

        if initial_message:
            await session.prompt(
                initial_message,
                PromptOptions(images=initial_images),
            )

        for message in messages:
            await session.prompt(message)

        await session.wait_for_idle()

        if mode == "text":
            state = session.state
            last_message = state.messages[-1] if state.messages else None
            if last_message and last_message.get("role") == "assistant":
                stop_reason = last_message.get("stopReason")
                if stop_reason in ("error", "aborted"):
                    raw_error = last_message.get("errorMessage") or f"Request {stop_reason}"
                    error_message = format_api_error_message(
                        str(raw_error),
                        provider=str(last_message.get("provider") or ""),
                        model_id=str(last_message.get("model") or ""),
                    )
                    print(error_message, file=sys.stderr)
                    exit_code = 1
                else:
                    for content in last_message.get("content", []):
                        if content.get("type") == "text":
                            write_raw_stdout(f"{content.get('text', '')}\n")

        return exit_code
    except Exception as error:
        print(str(error), file=sys.stderr)
        return 1
    finally:
        unregister_signal_handlers()
        await dispose_runtime()
        await flush_raw_stdout()
